src/Tools/GenerateDocumentationAndConfigFiles/Program.cs

```diff
@@ -303,17 +303,17 @@ string getDisableNetAnalyzersImport()
                 }
             }
 
-            string getCodeAnalysisTreatWarningsNotAsErrors()
+            string getCodeAnalysisTreatWarningsAsErrors()
             {
                 var allRuleIds = string.Join(';', allRulesById.Keys);
                 return $@"
   <!-- 
-    This property group prevents the rule ids implemented in this package to be bumped to errors when
-    the 'CodeAnalysisTreatWarningsAsErrors' = 'false'.
+    This property group handles 'CodeAnalysisTreatWarningsAsErrors' for the CA rule ids implemented in this package.
   -->
   <PropertyGroup>
     <CodeAnalysisRuleIds>{allRuleIds}</CodeAnalysisRuleIds>
     <WarningsNotAsErrors Condition=""'$(CodeAnalysisTreatWarningsAsErrors)' == 'false'"">$(WarningsNotAsErrors);$(CodeAnalysisRuleIds)</WarningsNotAsErrors>
+    <WarningsAsErrors Condition=""'$(CodeAnalysisTreatWarningsAsErrors)' == 'true' and '$(TreatWarningsAsErrors)' != 'true'"">$(WarningsAsErrors);$(CodeAnalysisRuleIds)</WarningsAsErrors>
```

Comment:
If `$(TreatWarningsAsErrors)` is not defined but `$(CodeAnalysisTreatWarningsAsErrors)` is set to truw this condition will occur is that correct?

Reply:
I guess I need you to write out a TreatWarningsAsErrors.md file where you lay out the spec here. From my understanding we have:

- `CodeAnalysisTreatWarningsAsErrors`
- `TreatWarningsAsErrors`
- `WarningsAsErrors`
- `WarningsNotAsErrors`

Without a spec laying out what the expected behavior of all these is I do not know how to review this.

Reply:
@jmarolf Following are the precedence rules as per the values of these properties.

1. For non-CAxxxx rules
   1. `CodeAnalysisTreatWarningsAsErrors` has no relevance and is ignored.
   2. `TreatWarningsAsErrors`, if not set, defaults to false. If true, this property translates to `/warnaserror` command line switch.
   3. Compiler bumps all warnings to errors iff `TreatWarningsAsErrors` is true. Users can prevent escalation or force escalation of individual warnings to errors by appending the IDs to `WarningsNotAsErrors` or `WarningsAsErrors`, which just translate to `/warnaserror[+|-]:<%rule ids%> on the command line.

2. For CAxxxx rules:
   1. If `CodeAnalysisTreatWarningsAsErrors` and `TreatWarningsAsErrors` both are not set, no bulk settings to escalate or de-escalate warnings to errors is done.
   2. If `CodeAnalysisTreatWarningsAsErrors` is set, it overrides `TreatWarningsAsErrors` to determine if CA warnings are bulk escalated to errors or not.
   3. If `CodeAnalysisTreatWarningsAsErrors` is not set, it defaults to `TreatWarningsAsErrors`.
   4. If final value of `CodeAnalysisTreatWarningsAsErrors = false`, we append all CA rule IDs to `WarningsNotAsErrors` to ensure they are not escalated to errors. Users can still bump individual rule IDs to errors by editorconfig/ruleset entry, etc.
   5. If final value of `CodeAnalysisTreatWarningsAsErrors = true`, we append all CA rule IDs to `WarningsAsErrors` to ensure they are escalated to errors. This is the change done in this PR. I have tried to optimize it a bit more by avoiding this append if `TreatWarningsAsErrors` is also true, because then the compiler itself will take care of bumping all warnings to errors, and we don't need to pollute the command line with large number of CA rules IDs in a `/warnaserror+` switch. I expect this to be the most common case as well (`TreatWarningsAsErrors` is set by user to true, `CodeAnalysisTreatWarningsAsErrors` is never set and hence defaults to `true`), and I want to ensure we don't end up polluting the entire command line with CA rules IDs unless `TreatWarningsAsErrors` and `CodeAnalysisTreatWarningsAsErrors` have different settings.